module/localisation/Localisation/src/Localisation.cpp

```diff
@@ -0,0 +1,432 @@
+            || observation.y() >= fieldline_map.map.cols()) {
+            return -1;
+        }
+        else {
+            return fieldline_map.map(observation.x(), observation.y());
+        }
+    }
+
+    Eigen::Vector2i Localisation::observation_relative(const Eigen::Matrix<double, 3, 1> particle,
+                                                       const Eigen::Vector2d observation) {
+        // Calculate the position of observation relative to the field [m]
+        double c       = cos(particle(2));
+        double s       = sin(particle(2));
+        double x_field = observation(0) * c - observation(1) * s + particle(0);
+        double y_field = observation(0) * s + observation(1) * c + particle(1);
+
+        // Get the associated position in the map [x, y]
+        int x_map = fieldline_map.map.rows() / 2 - std::round(y_field / cfg.grid_size);
+        int y_map = fieldline_map.map.cols() / 2 + std::round(x_field / cfg.grid_size);
+
+        return Eigen::Vector2i(x_map, y_map);
+    }
+
+    double Localisation::calculate_weight(const Eigen::Matrix<double, 3, 1> particle,
+                                          const std::vector<Eigen::Vector2d>& observations) {
+        double weight         = 0;
+        double n_observations = observations.size();
+
+        for (auto observation : observations) {
+            // Get the position of the particle in the map to check if robot is on the field [x, y]
+            Eigen::Vector2i particle_position = observation_relative(particle, Eigen::Vector2d(0.0, 0.0));
+            double particle_occupancy         = get_occupancy(particle_position);
+
+            // Get the position of the observation in the map for this particle [x, y]
+            Eigen::Vector2i map_position = observation_relative(particle, observation);
+
+            // Get the occupancy value of the observation at this position in the map
+            double observation_occupancy = get_occupancy(map_position);
+
+            // If the particle_occupancy is -1 then the robot is outside the map, penalise the particle
+            if (particle_occupancy == -1) {
+                weight = 0;
+            }
+            // If the observation_occupancy is -1 then the observation is outside the map, penalise the particle.
+            // Reduce the weight of the particle by the 1 / number of observations
+            else if (observation_occupancy == -1) {
+                weight -= (1.0 / n_observations);
+            }
+            else if (observation_occupancy > 0) {
+                // Increase the weight of the particle by the occupancy/n_observations, such that the maximum weight is
+                // between [0, 1]
+                weight += observation_occupancy / n_observations;
+            }
+        }
+
+        // Ensure the weight is positive
+        weight = std::max(weight, 0.0);
+
+        return weight;
+    }
+
+    Eigen::Vector3d Localisation::compute_mean() {
```

Comment:
I feel like a lot of these functions could be utilities.

Reply:
I agree, but at that point, we may as well move implementation over to use a general filter implementation within utility such as the current ParticleFilter.hpp, but I'd say thats out of scope of this PR atm.